File: cpp/src/engine.cpp
#include "frtb/engine.hpp"

#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <limits>
#include <sstream>
#include <stdexcept>

#include "frtb/json.hpp"

namespace frtb {

const std::vector<double>& PnlTable::series(const std::string& col) const {
    auto it = data.find(col);
    if (it == data.end())
        throw std::invalid_argument("PnlTable: no column '" + col + "'");
    return it->second;
}

namespace {

std::vector<std::string> split_csv(const std::string& line) {
    std::vector<std::string> out;
    std::string cell;
    std::istringstream ss(line);
    while (std::getline(ss, cell, ',')) out.push_back(cell);
    if (!line.empty() && line.back() == ',') out.emplace_back();
    return out;
}

std::string strip(const std::string& s) {
    std::size_t a = s.find_first_not_of(" \t\r\n");
    if (a == std::string::npos) return "";
    std::size_t b = s.find_last_not_of(" \t\r\n");
    return s.substr(a, b - a + 1);
}

}  // namespace

PnlTable load_pnl_csv(const std::string& path) {
    std::ifstream f(path);
    if (!f) throw std::invalid_argument("load_pnl_csv: cannot open " + path);
    std::string line;
    if (!std::getline(f, line))
        throw std::invalid_argument("load_pnl_csv: " + path + " is empty");
    auto header = split_csv(strip(line));
    PnlTable table;
    int date_idx = -1;
    for (std::size_t i = 0; i < header.size(); ++i) {
        std::string h = strip(header[i]);
        if (h == "date")
            date_idx = static_cast<int>(i);
        else
            table.columns.push_back(h);
    }
    if (date_idx < 0)
        throw std::invalid_argument("load_pnl_csv: " + path + " must have a 'date' column");
    for (const std::string& c : table.columns) table.data[c] = {};
    while (std::getline(f, line)) {
        if (strip(line).empty()) continue;
        auto cells = split_csv(line);
        cells.resize(header.size());
        std::size_t k = 0;
        for (std::size_t i = 0; i < header.size(); ++i) {
            if (static_cast<int>(i) == date_idx) continue;
            std::string cell = strip(cells[i]);
            double v;
            if (cell.empty()) {
                v = std::numeric_limits<double>::quiet_NaN();
            } else {
                char* end = nullptr;
                v = std::strtod(cell.c_str(), &end);
                if (end != cell.c_str() + cell.size())
                    throw std::invalid_argument("load_pnl_csv: bad number '" + cell + "' in " +
                                                path);
            }
            table.data[table.columns[k]].push_back(v);
            ++k;
        }
    }
    if (table.columns.empty() || table.data[table.columns.front()].empty())
        throw std::invalid_argument("load_pnl_csv: " + path + " contains no data rows");
    return table;
}

CategoryPnl desk_categories(const std::string& desk, const PnlTable& hypo) {
    const std::string prefix = desk + "_";
    CategoryPnl out;
    for (const std::string& c : hypo.columns)
        if (c.rfind(prefix, 0) == 0) out.emplace_back(c.substr(prefix.size()), hypo.series(c));
    return out;
}

Results compute_results(const std::string& data_dir) {
    Results res;
    res.params = load_params(data_dir + "/sbm_params.json");
    res.market = load_market(data_dir + "/curves.csv", data_dir + "/spots.csv");
    res.desks = load_portfolio(data_dir + "/portfolio.json");
    PnlTable hypo = load_pnl_csv(data_dir + "/pnl_hypo.csv");
    PnlTable rtpl = load_pnl_csv(data_dir + "/pnl_rtpl.csv");
    PnlTable var99 = load_pnl_csv(data_dir + "/pnl_var.csv");
    std::vector<NmrfEntry> nmrf;
    {
        json::Value doc = json::parse_file(data_dir + "/nmrf.json");
        for (const json::Value& e : doc.at("factors").array)
            nmrf.push_back({e.at("factor").as_string(), e.at("desk").as_string(),
                            e.at("stressed_loss").as_number()});
    }

    std::vector<std::string> desk_names;  // std::map keys are already sorted
    for (const auto& kv : res.desks) desk_names.push_back(kv.first);
    std::vector<Instrument> all_instruments;
    for (const std::string& d : desk_names)
        for (const Instrument& i : res.desks.at(d).instruments) all_instruments.push_back(i);

    // ---- SA per desk + firm ----------------------------------------------
    for (const std::string& d : desk_names)
        res.sa[d] = compute_sa(res.desks.at(d).instruments, res.market, res.params);
    res.sa["firm"] = compute_sa(all_instruments, res.market, res.params);
    res.sens_firm = compute_sensitivities(all_instruments, res.market, res.params);

    // ---- IMA per desk -----------------------------------------------------
    for (const std::string& d : desk_names) {
        CategoryPnl cats = desk_categories(d, hypo);
        if (cats.empty())
            throw std::invalid_argument("compute_results: no category P&L columns for desk '" +
                                        d + "'");
        const std::vector<double>& full = hypo.series(d);
        ImaDeskResult r;
        r.es_base = es_base_10d(full, res.params.ima_alpha);
        r.es_lh = es_lh_scaled(full, cats, res.params.category_lh, res.params.lh_ladder,
                               res.params.ima_alpha);
        r.imcc = imcc(full, cats, res.params);
        r.backtest = backtest(full, var99.series(d), res.params);
        r.plat = plat_test(full, rtpl.series(d), res.params);
        std::vector<NmrfEntry> desk_nmrf;
        for (const NmrfEntry& e : nmrf)
            if (e.desk == d) desk_nmrf.push_back(e);
        r.ses = ses(desk_nmrf);
        r.capital_core = ima_capital(r.imcc, r.backtest.multiplier, r.ses);
        r.plat_surcharge =
            plat_surcharge(r.plat.zone, res.sa.at(d).capital(), r.capital_core, res.params);
        r.capital = r.capital_core + r.plat_surcharge;
        res.ima[d] = r;
    }

    // ---- validation checks ------------------------------------------------
    ValidationBlock& val = res.validation;
    val.benchmark_max_diff = benchmark_max_diff();
    val.sensitivity_max_diff = sensitivity_max_diff();
    val.stability_base_capital = res.sa.at("firm").sbm.capital;
    val.stability_capital_rw_up10 =
        sbm_capital(res.sens_firm, res.market, res.params.with_girr_delta_rw_scaled(1.1)).capital;
    val.stability_capital_rw_dn10 =
        sbm_capital(res.sens_firm, res.market, res.params.with_girr_delta_rw_scaled(0.9)).capital;
    val.stability_rel_change =
        (val.stability_base_capital > 0.0)
            ? std::max(std::abs(val.stability_capital_rw_up10 - val.stability_base_capital),
                       std::abs(val.stability_capital_rw_dn10 - val.stability_base_capital)) /
                  val.stability_base_capital
            : 0.0;
    for (const std::string& d : desk_names) val.data_quality[d] = data_quality(hypo.series(d));

    for (const std::string& d : desk_names) {
        DeskCheckInputs inputs;
        inputs.benchmark_max_diff = val.benchmark_max_diff;
        inputs.sensitivity_max_diff = val.sensitivity_max_diff;
        inputs.stability_rel_change = val.stability_rel_change;
        inputs.backtest_zone = res.ima.at(d).backtest.zone;
        inputs.plat_zone = res.ima.at(d).plat.zone;
        inputs.stale_days = val.data_quality.at(d).stale_days;
        inputs.gaps = val.data_quality.at(d).gaps;
        val.findings[d] = classify_findings(inputs);
        val.verdicts[d] = overall_verdict(val.findings.at(d));
    }

    val.report_md = render_report(res);
    return res;
}

}  // namespace frtb

File: rust/src/engine.rs
//! End-to-end orchestration: load the bundled data set, compute SA (SBM +
//! DRC + RRAO), the IMA sketch (ES/IMCC, backtesting, PLAT, SES) and the
//! independent validation results for every desk and for the firm.
//!
//! Fully deterministic: pure revaluation and closed-form statistics, no
//! RNG.

use std::collections::BTreeMap;
use std::path::Path;

use serde_json::Value;

use crate::error::{invalid, FrtbError, Result};
use crate::ima::{backtest, es_base_10d, es_lh_scaled, ima_capital, imcc, ses, DeskIma, NmrfEntry};
use crate::instruments::{load_portfolio, Desk, Instrument};
use crate::market::{load_market, Market};
use crate::params::{load_params, SbmParams};
use crate::plat::{plat_surcharge, plat_test};
use crate::sa::{
    drc_charge, drc_positions_from_instruments, rrao_charge, sbm_capital, SbmResult,
};
use crate::sensitivities::{compute_sensitivities, Sensitivities};
use crate::validation::{
    benchmark_max_diff, classify_findings, data_quality, overall_verdict, render_report,
    sensitivity_max_diff, DeskCheckInputs, ValidationSummary,
};

/// P&L table: (column name, series) pairs in file column order (the order
/// matters — category sums replicate the reference's dict iteration).
pub type PnlTable = Vec<(String, Vec<f64>)>;

/// Load a P&L CSV (`date` + numeric columns) into a [`PnlTable`].
///
/// Empty cells become NaN (picked up by the data-quality check).
pub fn load_pnl_csv(path: &Path) -> Result<PnlTable> {
    let text = std::fs::read_to_string(path)
        .map_err(|e| FrtbError::Io(format!("cannot read {}: {e}", path.display())))?;
    let mut lines = text.lines().filter(|l| !l.trim().is_empty());
    let header: Vec<&str> = lines
        .next()
        .ok_or_else(|| FrtbError::Io(format!("{}: empty CSV", path.display())))?
        .split(',')
        .map(str::trim)
        .collect();
    let date_idx = header.iter().position(|h| *h == "date").ok_or_else(|| {
        FrtbError::Invalid(format!("load_pnl_csv: {} must have a 'date' column", path.display()))
    })?;
    let mut out: PnlTable = header
        .iter()
        .enumerate()
        .filter(|(i, _)| *i != date_idx)
        .map(|(_, h)| (h.to_string(), Vec::new()))
        .collect();
    for line in lines {
        let cells: Vec<&str> = line.split(',').map(str::trim).collect();
        let mut k = 0;
        for (i, cell) in cells.iter().enumerate() {
            if i == date_idx {
                continue;
            }
            let v = if cell.is_empty() {
                f64::NAN
            } else {
                cell.parse::<f64>().map_err(|_| {
                    FrtbError::Invalid(format!("load_pnl_csv: cannot parse '{cell}'"))
                })?
            };
            out[k].1.push(v);
            k += 1;
        }
    }
    if out.is_empty() || out[0].1.is_empty() {
        return invalid(format!("load_pnl_csv: {} contains no data rows", path.display()));
    }
    Ok(out)
}

/// Look up one column of a [`PnlTable`].
pub fn pnl_column<'a>(table: &'a PnlTable, name: &str) -> Result<&'a Vec<f64>> {
    table
        .iter()
        .find(|(c, _)| c == name)
        .map(|(_, v)| v)
        .ok_or_else(|| FrtbError::Invalid(format!("pnl table: missing column '{name}'")))
}

/// Extract the per-category P&L columns `<desk>_<cat>` for one desk,
/// preserving file column order.
pub fn desk_categories(desk: &str, table: &PnlTable) -> PnlTable {
    let prefix = format!("{desk}_");
    table
        .iter()
        .filter(|(c, _)| c.starts_with(&prefix))
        .map(|(c, v)| (c[prefix.len()..].to_string(), v.clone()))
        .collect()
}

/// SA results for one scope (a desk or the whole firm).
#[derive(Debug, Clone, PartialEq)]
pub struct SaScope {
    /// SBM drill-down and capital.
    pub sbm: SbmResult,
    /// DRC-lite charge.
    pub drc: f64,
    /// DRC hedge benefit ratio.
    pub drc_hbr: f64,
    /// Residual risk add-on.
    pub rrao: f64,
}

impl SaScope {
    /// SA capital = SBM + DRC + RRAO.
    pub fn capital(&self) -> f64 {
        self.sbm.capital + self.drc + self.rrao
    }
}

/// SA capital for one instrument scope: SBM + DRC-lite + RRAO.
pub fn compute_sa(
    instruments: &[Instrument],
    market: &Market,
    params: &SbmParams,
) -> Result<SaScope> {
    let sens = compute_sensitivities(instruments, market, params)?;
    let sbm = sbm_capital(&sens, market, params)?;
    let drc = drc_charge(&drc_positions_from_instruments(instruments, market)?, params)?;
    Ok(SaScope { sbm, drc: drc.charge, drc_hbr: drc.hbr, rrao: rrao_charge(instruments, params)? })
}

/// The full result tree computed from the bundled data directory.
#[derive(Debug, Clone, PartialEq)]
pub struct Results {
    /// The pinned parameter set.
    pub params: SbmParams,
    /// The loaded market snapshot.
    pub market: Market,
    /// Desks keyed by name.
    pub desks: BTreeMap<String, Desk>,
    /// SA per scope: every desk plus `"firm"`.
    pub sa: BTreeMap<String, SaScope>,
    /// Firm-wide sensitivities (reused by the stability check).
    pub sens_firm: Sensitivities,
    /// IMA sketch per desk.
    pub ima: BTreeMap<String, DeskIma>,
    /// Validation checks, findings, verdicts and the rendered report.
    pub validation: ValidationSummary,
}

/// Parse `nmrf.json` into entries.
fn load_nmrf(path: &Path) -> Result<Vec<NmrfEntry>> {
    let text = std::fs::read_to_string(path)
        .map_err(|e| FrtbError::Io(format!("cannot read {}: {e}", path.display())))?;
    let raw: Value = serde_json::from_str(&text)
        .map_err(|e| FrtbError::Io(format!("{}: bad JSON: {e}", path.display())))?;
    let factors = raw
        .get("factors")
        .and_then(Value::as_array)
        .ok_or_else(|| FrtbError::Invalid("nmrf.json: missing 'factors' list".into()))?;
    let mut out = Vec::new();
    for f in factors {
        out.push(NmrfEntry {
            factor: f
                .get("factor")
                .and_then(Value::as_str)
                .ok_or_else(|| FrtbError::Invalid("nmrf.json: entry missing 'factor'".into()))?
                .to_string(),
            desk: f
                .get("desk")
                .and_then(Value::as_str)
                .ok_or_else(|| FrtbError::Invalid("nmrf.json: entry missing 'desk'".into()))?
                .to_string(),
            stressed_loss: f
                .get("stressed_loss")
                .and_then(Value::as_f64)
                .ok_or_else(|| {
                    FrtbError::Invalid("nmrf.json: entry missing 'stressed_loss'".into())
                })?,
        });
    }
    Ok(out)
}

/// Compute the full result tree from the bundled data directory.
pub fn compute_results(data_dir: &Path) -> Result<Results> {
    let params = load_params(&data_dir.join("sbm_params.json"))?;
    let market = load_market(&data_dir.join("curves.csv"), &data_dir.join("spots.csv"))?;
    let desks = load_portfolio(&data_dir.join("portfolio.json"))?;
    let hypo = load_pnl_csv(&data_dir.join("pnl_hypo.csv"))?;
    let rtpl = load_pnl_csv(&data_dir.join("pnl_rtpl.csv"))?;
    let var99 = load_pnl_csv(&data_dir.join("pnl_var.csv"))?;
    let nmrf = load_nmrf(&data_dir.join("nmrf.json"))?;

    let desk_names: Vec<String> = desks.keys().cloned().collect();
    let all_instruments: Vec<Instrument> = desk_names
        .iter()
        .flat_map(|d| desks[d].instruments.iter().cloned())
        .collect();

    // ---- SA per desk + firm ---------------------------------------------
    let mut sa = BTreeMap::new();
    for d in &desk_names {
        sa.insert(d.clone(), compute_sa(&desks[d].instruments, &market, &params)?);
    }
    sa.insert("firm".to_string(), compute_sa(&all_instruments, &market, &params)?);
    let sens_firm = compute_sensitivities(&all_instruments, &market, &params)?;

    // ---- IMA per desk ----------------------------------------------------
    let mut ima = BTreeMap::new();
    for d in &desk_names {
        let cats = desk_categories(d, &hypo);
        if cats.is_empty() {
            return invalid(format!("compute_results: no category P&L columns for desk '{d}'"));
        }
        let full = pnl_column(&hypo, d)?;
        let es_b = es_base_10d(full, params.ima_alpha)?;
        let es_lh =
            es_lh_scaled(full, &cats, &params.category_lh, &params.lh_ladder, params.ima_alpha)?;
        let imcc_d = imcc(full, &cats, &params)?;
        let bt = backtest(full, pnl_column(&var99, d)?, &params)?;
        let pl = plat_test(full, pnl_column(&rtpl, d)?, &params)?;
        let desk_nmrf: Vec<NmrfEntry> =
            nmrf.iter().filter(|e| e.desk == *d).cloned().collect();
        let ses_d = ses(&desk_nmrf)?;
        let core = ima_capital(imcc_d, bt.multiplier, ses_d, 0.0)?;
        let surcharge = plat_surcharge(&pl.zone, sa[d].capital(), core, &params)?;
        ima.insert(
            d.clone(),
            DeskIma {
                es_base: es_b,
                es_lh,
                imcc: imcc_d,
                backtest: bt,
                plat: pl,
                ses: ses_d,
                capital_core: core,
                plat_surcharge: surcharge,
                capital: core + surcharge,
            },
        );
    }

    // ---- validation checks ----------------------------------------------
    let bench = benchmark_max_diff()?;
    let sens_diff = sensitivity_max_diff()?;
    let base_cap = sa["firm"].sbm.capital;
    let cap_up =
        sbm_capital(&sens_firm, &market, &params.with_girr_delta_rw_scaled(1.1)?)?.capital;
    let cap_dn =
        sbm_capital(&sens_firm, &market, &params.with_girr_delta_rw_scaled(0.9)?)?.capital;
    let stability_rel = if base_cap > 0.0 {
        (cap_up - base_cap).abs().max((cap_dn - base_cap).abs()) / base_cap
    } else {
        0.0
    };
    let mut dq = BTreeMap::new();
    for d in &desk_names {
        dq.insert(d.clone(), data_quality(pnl_column(&hypo, d)?)?);
    }

    let mut findings = BTreeMap::new();
    let mut verdicts = BTreeMap::new();
    for d in &desk_names {
        let inputs = DeskCheckInputs {
            benchmark_max_diff: bench,
            sensitivity_max_diff: sens_diff,
            stability_rel_change: stability_rel,
            backtest_zone: ima[d].backtest.zone.clone(),
            plat_zone: ima[d].plat.zone.clone(),
            stale_days: dq[d].stale_days,
            gaps: dq[d].gaps,
        };
        let f = classify_findings(&inputs);
        verdicts.insert(d.clone(), overall_verdict(&f).to_string());
        findings.insert(d.clone(), f);
    }

    let mut validation = ValidationSummary {
        benchmark_max_diff: bench,
        sensitivity_max_diff: sens_diff,
        stability_base_capital: base_cap,
        stability_capital_rw_up10: cap_up,
        stability_capital_rw_dn10: cap_dn,
        stability_rel_change: stability_rel,
        data_quality: dq,
        findings,
        verdicts,
        report_md: String::new(),
    };
    validation.report_md = render_report(&ima, &validation);

    Ok(Results { params, market, desks, sa, sens_firm, ima, validation })
}
